drgn_tools/cpuinfo.py

```diff
@@ -0,0 +1,581 @@
+
+    return mitigation
+
+
+def get_mds_mitigation(prog: Program, cpu_caps_bugs: Object) -> str:
+    """
+    Extracts Mitigation for MDS
+    """
+    mitigation = (
+        prog["mds_strings"][prog["mds_mitigation"]].string_().decode("utf-8")
+    )
+    mitigation += "; SMT "
+    ncapints = len(prog["x86_cap_flags"]) // 32
+    if test_bit(X86_FEATURE_HYPERVISOR, cpu_caps_bugs):
+        mitigation += "Host state unknown"
+    elif test_bit(ncapints * 32 + X86_BUG_MSBDS_ONLY, cpu_caps_bugs):
+        if prog["mds_mitigation"] == prog.constant("MDS_MITIGATION_OFF"):
+            mitigation += "vulnerable"
+        elif check_smt_enabled(prog):
+            mitigation += "mitigated"
+        else:
+            mitigation += "disabled"
+    else:
+        if check_smt_enabled(prog):
+            mitigation += "vulnerable"
+        else:
+            mitigation += "disabled"
+
+    return mitigation
+
+
+def get_taa_mitigation(prog: Program, cpu_caps_bugs: Object) -> str:
+    """
+    Extracts Mitigation for tsx_async_abort
+    """
+    mitigation = (
+        prog["taa_strings"][prog["taa_mitigation"]].string_().decode("utf-8")
+    )
+
+    if prog["taa_mitigation"] == prog.constant(
+        "TAA_MITIGATION_TSX_DISABLED"
+    ) or prog["taa_mitigation"] == prog.constant("TAA_MITIGATION_OFF"):
+        pass
+    elif test_bit(X86_FEATURE_HYPERVISOR, cpu_caps_bugs):
+        mitigation += "; SMT Host state unknown"
+    else:
+        mitigation += "; SMT "
+        if check_smt_enabled(prog):
+            mitigation += "vulnerable"
+        else:
+            mitigation += "disabled"
+
+    return mitigation
+
+
+def get_itlb_multihit_mitigation(prog: Program, cpu_caps_bugs: Object) -> str:
+    """
+    Extracts Mitigation for itlb_multihit
+    """
+    if "l1tf_vmx_states" not in prog:
+        mitigation = "Processor vulnerable"
+    elif "cr4_read_shadow" in prog:
+        if (not test_bit(X86_FEATURE_MSR_IA32_FEAT_CTL, cpu_caps_bugs)) or (
+            not test_bit(X86_FEATURE_VMX, cpu_caps_bugs)
+        ):
+            mitigation = "KVM: Mitigation: VMX unsupported"
+        elif not per_cpu(prog["cpu_tlbstate"], 0).cr4 & X86_CR4_VMXE:
+            mitigation = "KVM: Mitigation: VMX disabled"
+    elif prog["itlb_multihit_kvm_mitigation"]:
+        mitigation = "KVM: Mitigation: Split huge pages"
+    else:
+        mitigation = "KVM: Vulnerable"
+
+    return mitigation
+
+
+def get_srbds_mitigation(prog: Program, cpu_caps_bugs: Object) -> str:
+    """
+    Extracts Mitigation for SRBDS
+    """
+    return (
+        prog["srbds_strings"][prog["srbds_mitigation"]]
+        .string_()
+        .decode("utf-8")
+    )
+
+
+def get_mmio_stale_data_mitigation(
+    prog: Program, cpu_caps_bugs: Object
+) -> str:
+    """
+    Extracts Mitigation for mmio_stale_data and mmio_unknown
+    """
+    ncapints = len(prog["x86_cap_flags"]) // 32
+    if test_bit(ncapints * 32 + X86_BUG_MMIO_UNKNOWN):
```

Comment:
I guess when you write it like that, there's no reason to divide by 30 :)
Also, I think you forgot to add `cpu_caps_bugs` to the argument.
```suggestion
    num_caps = len(prog["x86_cap_flags"])
    if test_bit(num_caps + X86_BUG_MMIO_UNKNOWN, cpu_caps_bugs):
```

Reply:
The division is actually needed, because each bit in `cpu_caps_bugs` corresponds to an element of `x86_cap_flags`, so it has `num_caps * 32` elements.